### PR TITLE
Microwave Overhaul, or: death to user << browse

### LINKED ISSUE #6728: Microwave Duplication
The med bay break room microwave (if that makes any difference at all) duplicated a Donk-pocket.

I put a pocket into the microwave and turned the microwave on. I waited a bit and got impatient and ejected the contents. The Donk-pocket ejected and I ate it. The microwave then dinged and another Donk-pocket was ejected.

### PR DESCRIPTION
**Pull request by JohnWildkins**
The microwave code is old enough to be starting fifth grade this year. Many UI frameworks have come and gone since its construction, and procs such as proc/wzhzhzh confound many a coder to this day. As I dream of a 21st century kitchen, I think of a better tomorrow, of UIs made after the development of any form of web standard.

With that, I bring you the Microwave Overhaul. What's done so far (to be updated):

- Anything to do with the HTML-based UI has been completely gutted. In its place, a Vue UI that hopefully looks the part and functions to exact feature parity with the old UI.
- Eject() verb, the cause of more than one issue, has been removed in place of a new Abort function.
- Abort() and Stop() have been merged as their usage was essentially identical.
- Individual non-reagent ingredients can now be ejected at-will.
- Reagents can now be ejected into a held container.
- Microwave cook time is now displayed as a progress bar on the UI.
- Reagent containers now proc UI updates when pouring onto an object with active UIs (code reviewers, please, if you can find a better way to accomplish what I'm doing here, please do. I hate to do this)
- proc/wzhzhzh now has an actual name befitting its purpose. A memorial comment was placed in remembrance.

What still needs to be done:

- [x] Microwave upgrade parts (per #7174) should have a benefit (cook time, power usage, etc.).
- [x] More rigorous bug testing, especially of edge cases w/ the UI, emptying into beakers, etc.
- [x] Look into better methods for both catching reagent adding and emptying reagents into beakers
- [x] UX tweaks to the microwave interface per feedback
- [x] Tearing apart the main cook() loop in-depth and making it make any kind of actual sense.
- [x] Murdering useless procs, replacing them with slightly less useless ones.
- [x] Refactor cook() again so it's actually sensible - fixes #6728.

**Previews:**

### CODE REVIEW

**Comment by JohnWildkins** [comment]
!wip

File: code/game/machinery/kitchen/microwave.dm [hunk 1]
```diff
@@ -253,38 +231,42 @@
...
+				stop()
 				return
 			broke()
 			cooked = fail()
 			cooked.forceMove(src.loc)
 			return
 		else
-			if (!wzhzhzh(40))
-				abort()
+			if (!cook_for_time(40))
```
Comment (NonQueueingMatt) [inline code/game/machinery/kitchen/microwave.dm]: These are a lot of magic numbers. They should really be a define.

File: code/game/machinery/kitchen/microwave.dm [hunk 2]
```diff
@@ -253,38 +231,42 @@
...
 	else
-		var/halftime = round((recipe.time*4)/10/2)
-		if (!wzhzhzh(halftime))
-			abort()
-			return
-		if (!wzhzhzh(halftime))
-			abort()
+		var/cut_time = round((recipe.time*4)/10)
+		// if (!cook_for_time(halftime))
+		// 	abort()
```
Comment (NonQueueingMatt) [inline code/game/machinery/kitchen/microwave.dm]: Remove, don't comment.

File: code/game/machinery/kitchen/microwave.dm [hunk 3]
```diff
@@ -324,16 +306,21 @@
...
-		dispose(0) //clear out anything left
+		eject(0) //clear out anything left
 		stop()
 
 		return
 
-/obj/machinery/microwave/proc/wzhzhzh(var/seconds as num) // Whoever named this proc is fucking literally Satan. ~ Z
-	for (var/i=1 to seconds)
-		if (stat & (NOPOWER|BROKEN))
+/obj/machinery/microwave/proc/cook_for_time(var/cook_time as num) // Formerly proc/wzhzhzh, may it rest in peace
```
Comment (NonQueueingMatt) [inline code/game/machinery/kitchen/microwave.dm]:
```suggestion
/obj/machinery/microwave/proc/cook_for_time(var/cook_time) // Formerly proc/wzhzhzh, may it rest in peace
```

File: code/game/machinery/kitchen/microwave.dm [hunk 4]
```diff
@@ -416,40 +401,47 @@
...
+						RC.reagents.add_reagent(R.id, R.volume)
+						src.reagents.remove_reagent(R.id, R.volume)
+					else
+						to_chat(usr, "<span class='notice'>You empty [vol_diff] units of [R.name] into your [RC.name].</span>")
+						RC.reagents.add_reagent(R.id, vol_diff)
+						src.reagents.remove_reagent(R.id, vol_diff)
+					SSvueui.check_uis_for_change(src)
+					break
+				else
+					to_chat(usr, "<span class='warning'>You need to be holding a valid container to empty [R.name]!</span>")
```
Comment (NonQueueingMatt) [inline code/game/machinery/kitchen/microwave.dm]: I mean... how are they gonna empty something that isnt a container?

Comment (JohnWildkins) [reply code/game/machinery/kitchen/microwave.dm]: It's an error message for trying to empty reagents out of the microwave without a container in their hand, not for trying to empty not-a-container into the microwave.

File: code/game/machinery/kitchen/microwave.dm (hunk 2 above)
Comment (JohnWildkins) [reply code/game/machinery/kitchen/microwave.dm]: sorted in next commit

File: code/game/machinery/kitchen/microwave.dm (hunk 1 above)
Comment (JohnWildkins) [reply code/game/machinery/kitchen/microwave.dm]: sorted (probably) in next commit

File: code/game/machinery/kitchen/microwave.dm (hunk 3 above)
Comment (JohnWildkins) [reply code/game/machinery/kitchen/microwave.dm]: sorted in next commit

File: code/game/machinery/kitchen/microwave.dm [hunk 5]
```diff
@@ -97,141 +109,113 @@
...
-
-	user << browse("<HEAD><TITLE>Microwave Controls</TITLE></HEAD><TT>[dat]</TT>", "window=microwave")
-	onclose(user, "microwave")
-	return
+			newdata["cookingreas"] += list(list("name" = "[R.name]", "amt" = "[R.volume]"))
 
+	if(newdata["cookingobjs"] && contents && newdata["cookingobjs"].len != contents.len)
+		. = newdata
+	if(newdata["cookingreas"] && reagents.reagent_list && newdata["cookingreas"].len != reagents.reagent_list.len)
+		. = newdata
```
Comment (Karolis2011) [inline code/game/machinery/kitchen/microwave.dm]: What is this mess?

File: code/game/machinery/kitchen/microwave.dm [hunk 6]
```diff
@@ -97,141 +109,113 @@
...
-				dat += {"<B>[capitalize(O)]:</B> [N] [lowertext(O)]\s<BR>"}
-			else
-				if (N==1)
-					dat += {"<B>[capitalize(O)]:</B> [N] [items_measures[O]]<BR>"}
-				else
-					dat += {"<B>[capitalize(O)]:</B> [N] [items_measures_p[O]]<BR>"}
+VUEUI_MONITOR_VARS(/obj/machinery/microwave, microwavemonitor)
+	watch_var("operating", "on", CALLBACK(null, .proc/transform_to_boolean, FALSE))
 
+/obj/machinery/microwave/vueui_data_change(var/list/newdata, var/mob/user, var/datum/vueui/ui)
```
Comment (Karolis2011) [inline code/game/machinery/kitchen/microwave.dm]: `newdata` is kinda obsolete name for this argument as currently it's being called more often `data` it's name represents that data is incomming change from ui. Reminder that if `ui.data == newdata` then it's just check for change.

File: code/game/machinery/kitchen/microwave.dm [hunk 7]
```diff
@@ -97,141 +109,113 @@
...
+	if(monitordata)
+		. = newdata = monitordata
+
+	newdata["cookingobjs"] = list()
+	newdata["cookingreas"] = list()
+
+	VUEUI_SET_CHECK_IFNOTSET(newdata["cook_time"], 40, ., newdata)
+	VUEUI_SET_CHECK_IFNOTSET(newdata["cur_time"], 0, ., newdata)
+
+	if (contents && contents.len)
```
Comment (Karolis2011) [inline code/game/machinery/kitchen/microwave.dm]: `LAZYLEN(contents)` would look way nicer

File: code/game/machinery/kitchen/microwave.dm [hunk 8]
```diff
@@ -97,141 +109,113 @@
...
+VUEUI_MONITOR_VARS(/obj/machinery/microwave, microwavemonitor)
+	watch_var("operating", "on", CALLBACK(null, .proc/transform_to_boolean, FALSE))
 
+/obj/machinery/microwave/vueui_data_change(var/list/newdata, var/mob/user, var/datum/vueui/ui)
+	var/monitordata = ..()
+	if(monitordata)
+		. = newdata = monitordata
+
+	newdata["cookingobjs"] = list()
+	newdata["cookingreas"] = list()
```
Comment (Karolis2011) [inline code/game/machinery/kitchen/microwave.dm]: Use `LAZYINITLIST(list)` to not override existing data.

File: code/game/machinery/kitchen/microwave.dm [hunk 9]
```diff
@@ -97,141 +109,113 @@
...
+
+	VUEUI_SET_CHECK_IFNOTSET(newdata["cook_time"], 40, ., newdata)
+	VUEUI_SET_CHECK_IFNOTSET(newdata["cur_time"], 0, ., newdata)
+
+	if (contents && contents.len)
+		var/list/cook_count = list()
+		for (var/obj/O in contents)
+			cook_count[O.name]++
+		for (var/C in cook_count)
+			newdata["cookingobjs"] += list(list("name" = "[C]", "qty" = "[cook_count[C]]"))
```
Comment (Karolis2011) [inline code/game/machinery/kitchen/microwave.dm]:
```suggestion
			VUEUI_SET_CHECK(newdata["cookingobjs"][C], cook_count[C], ., newdata)
```
This would prevent the mess that is below.

File: code/game/machinery/kitchen/microwave.dm [hunk 10]
```diff
@@ -97,141 +109,113 @@
...
-			dat = {"<b>Ingredients:</b><br>[dat]"}
-		dat += {"<HR><BR>\
-<A href='?src=\ref[src];action=cook'>Turn on!<BR>\
-<A href='?src=\ref[src];action=dispose'>Eject ingredients!<BR>\
-"}
-
-	user << browse("<HEAD><TITLE>Microwave Controls</TITLE></HEAD><TT>[dat]</TT>", "window=microwave")
-	onclose(user, "microwave")
-	return
+			newdata["cookingreas"] += list(list("name" = "[R.name]", "amt" = "[R.volume]"))
```
Comment (Karolis2011) [inline code/game/machinery/kitchen/microwave.dm]: If there are are duplicate reagents, this should be refactored to use `VUEUI_SET_CHECK` and maps (objects), instead of using arrays.

File: code/game/machinery/kitchen/microwave.dm [hunk 11]
```diff
@@ -324,18 +283,60 @@
...
+
+/obj/machinery/microwave/proc/update_cooking_ui(var/cook_time, var/cur_time)
+	for(var/datum/vueui/ui in SSvueui.get_open_uis(src))
+		ui.data["cook_time"] = cook_time
+		ui.data["cur_time"] = cur_time
+	SSvueui.check_uis_for_change(src)
+
+/obj/machinery/microwave/proc/close_cooking_uis()
+	for(var/datum/vueui/ui in SSvueui.get_open_uis(src))
+		ui.close()
```
Comment (Karolis2011) [inline code/game/machinery/kitchen/microwave.dm]: There is no point to this proc. `SSvueui.close_uis(obj)`

File: code/game/machinery/kitchen/microwave.dm [hunk 12]
```diff
@@ -324,18 +283,60 @@
...
+
+	stop(dirt, broke)
+
+	return TRUE
+
+/obj/machinery/microwave/proc/update_cooking_ui(var/cook_time, var/cur_time)
+	for(var/datum/vueui/ui in SSvueui.get_open_uis(src))
+		ui.data["cook_time"] = cook_time
+		ui.data["cur_time"] = cur_time
+	SSvueui.check_uis_for_change(src)
```
Comment (Karolis2011) [inline code/game/machinery/kitchen/microwave.dm]: You should not directly modify in this case ui data. Use var monitors.

In this case you should just keep track of when cooking started and when it should end. Use `world.time` DM and `wtime` state variable for time tracking. For how it can be used, look at faxmachine code.

File: code/game/machinery/kitchen/microwave.dm [hunk 13]
```diff
@@ -346,61 +347,38 @@
...
-	src.icon_state = "mwb" // Make it look all busted up and shit
-	src.visible_message("<span class='warning'>The microwave breaks!</span>") //Let them know they're stupid
-	src.broken = 2 // Make it broken so it can't be used util fixed
-	src.flags = null //So you can't add condiments
-	src.operating = 0 // Turn it off again aferwards
-	src.updateUsrDialog()
+	operating = FALSE // Turn it off again aferwards
+	if(muck || broken)
+		flags = null //So you can't add condiments
+		close_cooking_uis() // Can't use this anymore!
```
Comment (Karolis2011) [inline code/game/machinery/kitchen/microwave.dm]: This should not be needed as VueUI code should handle things that are broken and unusable via Topic states.

File: code/modules/reagents/reagent_containers.dm [hunk 14]
```diff
@@ -70,6 +70,8 @@
 	if(standard_splash_mob(user, target))
 		return
 	if(standard_pour_into(user, target))
+		if(SSvueui.get_open_uis(target)) // blame whoever decided beakers should go in afterattack()
+			SSvueui.check_uis_for_change(target)
```
Comment (Karolis2011) [inline code/modules/reagents/reagent_containers.dm]: Remove if, just do check for change. You are just wasting time.

File: vueui/src/components/view/cooking/microwave.vue [hunk 15]
```diff
@@ -0,0 +1,50 @@
+<template>
+  <div>
+    <template v-if="on == true">
+      <p>Cooking...</p>
+      <vui-progress v-if="cook_time" :value="cur_time" :max="cook_time"/>
+      <vui-button class="danger danger-control" icon="exclamation-triangle" push-state :params="{abort: 1}">Abort!</vui-button>
```
Comment (Karolis2011) [inline vueui/src/components/view/cooking/microwave.vue]: `push-state` is absolutely unnecessary here, as you do not read client side data or modify it on client side. This just adds bloat to network traffic.

File: vueui/src/components/view/cooking/microwave.vue [hunk 16]
```diff
@@ -0,0 +1,50 @@
+<template>
+  <div>
+    <template v-if="on == true">
```
Comment (Karolis2011) [inline vueui/src/components/view/cooking/microwave.vue]:
```suggestion
    <template v-if="on">
```

File: vueui/src/components/view/cooking/microwave.vue [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+      <p>Cooking...</p>
+      <vui-progress v-if="cook_time" :value="cur_time" :max="cook_time"/>
+      <vui-button class="danger danger-control" icon="exclamation-triangle" push-state :params="{abort: 1}">Abort!</vui-button>
+    </template>
+    <template v-else>
+      <h4>Ingredients</h4>
+      <ul>
+        <li v-if="cookingobjs.length == 0 && cookingreas.length == 0">The microwave is empty!</li>
```
Comment (Karolis2011) [inline vueui/src/components/view/cooking/microwave.vue]: Might want to use computed property, but I am still not sure as this is WIP.
Also migth want to make this statement not inside buletin list.
```suggestion
        <li v-if="!cookingobjs.length && !cookingreas.length">The microwave is empty!</li>
```

File: vueui/src/components/view/cooking/microwave.vue [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+    position: fixed;
+    right: 20px;
+    bottom: 20px;
+  }
+  .control {
+    position: fixed;
+    left: 20px;
+    bottom: 20px;
+  }
+</style>
```
Comment (Karolis2011) [inline vueui/src/components/view/cooking/microwave.vue]: Styles are a bit iffy, but this might need a better solution in future. Will suggest how to solve this in future.

File: code/game/machinery/kitchen/microwave.dm (hunk 5 above)
Comment (JohnWildkins) [reply code/game/machinery/kitchen/microwave.dm]: It's a relic from the initial copy I made of the fax code, the only real purpose it served currently was wiping the data if there was a length mismatch. I've replaced that now with a slightly less messy version.

File: code/game/machinery/kitchen/microwave.dm (hunk 11 above)
Comment (JohnWildkins) [reply code/game/machinery/kitchen/microwave.dm]: Good to know, thanks

File: code/game/machinery/kitchen/microwave.dm (hunk 12 above)
Comment (JohnWildkins) [reply code/game/machinery/kitchen/microwave.dm]: Didn't even think of that, and you're totally right. Hopefully the new version is a bit better.

File: code/modules/reagents/reagent_containers.dm (hunk 14 above)
Comment (JohnWildkins) [reply code/modules/reagents/reagent_containers.dm]: Fair enough, wasn't sure if one was cheaper than the other.

File: vueui/src/components/view/cooking/microwave.vue (hunk 18 above)
Comment (JohnWildkins) [reply vueui/src/components/view/cooking/microwave.vue]: Yeah, it's by no means perfect but I didn't want to mess around too much with the front-end until I was sure the back-end was solid

File: code/game/machinery/kitchen/microwave.dm [hunk 19]
```diff
@@ -185,31 +189,36 @@
...
+
+	// init data objects if they aren't already
+	LAZYINITLIST(data["cookingobjs"])
+	LAZYINITLIST(data["cookingreas"])
 
-	newdata["cookingobjs"] = list()
-	newdata["cookingreas"] = list()
+	// set up the progress bar vars
+	VUEUI_SET_CHECK(data["cook_start"], start_time, ., data)
+	VUEUI_SET_CHECK(data["cook_elapsed"], world.time, ., data) // needed to update progress
```
[the remark this answers is not in the document]
Comment (JohnWildkins) [reply code/game/machinery/kitchen/microwave.dm]: see I struggled with wtime for like thirty minutes and couldn't get it to work with the progress bar, and it turns out it's because I can't read and you need to do  `$root.$data.wtime` as you said below so honestly committing that there was a cry for help and I'm glad you came with deliverance

**Comment by Karolis2011** [comment]
Please also note, `world.time` and `wtime` is in deciseconds. Aka 20 in wtime is 2s.

Also Client side `$root.$data.wtime` is tracked on client side, and it means that progress bar will progress smootly.

**Comment by JohnWildkins** [comment]
Alright. Barring any other issues I cause myself, I think that's it for me adding to the UI side of things. Now I'll just be waiting for #7174 so I can rebase and add the upgrade parts, and I think that's it feature-wise. Going to give it another look-over tonight when I wake up, though.

File: code/game/machinery/kitchen/microwave.dm [hunk 20]
```diff
@@ -182,17 +183,31 @@
...
+	else if(dirty > 25)
+		to_chat(user, "It's a bit dirty.")
+
 /*******************
 *   Microwave Menu
 ********************/
 
 VUEUI_MONITOR_VARS(/obj/machinery/microwave, microwavemonitor)
 	watch_var("operating", "on", CALLBACK(null, .proc/transform_to_boolean, FALSE))
+	watch_var("recipe", "recipe")
```
Comment (Karolis2011) [inline code/game/machinery/kitchen/microwave.dm]: You need to use too boolean transform here.

Comment (JohnWildkins) [reply code/game/machinery/kitchen/microwave.dm]: Yeah, you're right. I had removed it as a test and it seemed to fix the issue, but that was just a placebo it seems like and my issue is actually `recipe` not being sent to state before the progress bar area is drawn.

**Comment by JohnWildkins** [comment]
Rebase successful with minor spaghetti. Now just to make part upgrades actually matter, find a way to properly black-hole upgrade parts and still give them back on deconstruction, and we should be more or less golden.

**Comment by JohnWildkins** [comment]
okay well either I've murdered travis with the worst code on earth or it's just broken but I think this is ready to ship more or less

**Comment by JohnWildkins** [comment]
!review

**Comment by alsoandanswer** [comment]
!review

File: code/game/machinery/kitchen/microwave.dm [hunk 21]
```diff
@@ -42,6 +59,25 @@
...
+		var/count = component_types[type]
+		if (count > 1)
+			for (var/i in 1 to count)
+				var/obj/t = new type
+				component_parts += t
+				t.forceMove(null)
+		else
+			var/obj/t = new type
+			component_parts += t
+			t.forceMove(null)
```
Comment (skull132) [inline code/game/machinery/kitchen/microwave.dm]: Remove if the if-clause, it'll work as intended.

File: code/game/machinery/kitchen/microwave.dm [hunk 22]
```diff
@@ -42,6 +59,20 @@
...
 		setup_recipes()
 
+	if(component_parts)
+		component_parts.Cut()
+	else
+		component_parts = list()
+
+	for (var/type in component_types)
+		var/count = component_types[type]
+		for(var/i in 1 to count)
```
Comment (Karolis2011) [inline code/game/machinery/kitchen/microwave.dm]: This runtimes. Fix it.

File: code/game/machinery/kitchen/microwave.dm [hunk 23]
```diff
@@ -167,86 +183,98 @@
...
+
+	// init data objects if they aren't already
+	LAZYINITLIST(data["cookingobjs"])
+	LAZYINITLIST(data["cookingreas"])
+
+	// if BYOND lists are smaller than UI, then something (or everything) was removed - wipe the list
+	if(LAZYLEN(contents) < LAZYLEN(data["cookingobjs"]))
+		VUEUI_SET_CHECK(data["cookingobjs"], list(), ., data)
+	if(LAZYLEN(reagents.reagent_list) < LAZYLEN(data["cookingreas"]))
+		VUEUI_SET_CHECK(data["cookingreas"], list(), ., data)
```
Comment (Karolis2011) [inline code/game/machinery/kitchen/microwave.dm]: Maybe you should directly alter UI data when removing ingredients? 
by doing something like:
```dm
for(var/i in SSvueui.get_open_uis(src))
    var/datum/vueui/ui = i
    ui.data["cookingobjs"] -= "[rmeovedobj]"

File: code/game/machinery/kitchen/microwave.dm [hunk 24]
```diff
@@ -255,167 +283,146 @@
...
-	src.visible_message("<span class='notice'>The microwave turns on.</span>", "<span class='notice'>You hear a microwave.</span>")
-	src.operating = 1
-	src.icon_state = "mw1"
-	src.updateUsrDialog()
+	start_time = world.time
+	end_time = cook_time + start_time
+	operating = TRUE
+
+	START_PROCESSING(SSprocessing, src)
+	addtimer(CALLBACK(src, .proc/half_time_process), cook_time / 2)
```
Comment (Karolis2011) [inline code/game/machinery/kitchen/microwave.dm]: If microwave is broken mid cooking timer just does it's thing anyway. Better migth want to add way to stop it or handle half time events inside `process()`

File: vueui/src/components/view/cooking/microwave.vue [hunk 25]
```diff
@@ -0,0 +1,69 @@
+<template>
+  <div>
+    <template v-if="on">
+      <p>{{ cookingMessage }}</p>
```
Comment (Karolis2011) [inline vueui/src/components/view/cooking/microwave.vue]: I don't like use of computed property to affect message. `pastHalfTime` could be replaced with `$root.$data.wtime > (start_time + (cook_time / 2))`
```suggestion
      <template v-if="pastHalfTime">
        <p v-if="!ingredientsPresent">There's nothing inside. Microwave works, though.</p>
        <p v-else-if="failed">Something doesn't look right...</p>
        <p v-else>It's cooking nicely!</p>
      </template>
      <p v-else>It's starting to cook...</p>
```

File: vueui/src/components/view/cooking/microwave.vue [hunk 26]
```diff
@@ -0,0 +1,69 @@
...
+      <vui-progress :value="$root.$data.wtime" :min="start_time" :max="start_time + cook_time"/>
```
Comment (Karolis2011) [inline vueui/src/components/view/cooking/microwave.vue]: progress component does not cap `value` value, so you need to add cap here or at progress component. Else it migth extend past it's limits. Like how vote timer counts a bit to negative numbers.

File: vueui/src/components/view/cooking/microwave.vue [hunk 27]
```diff
@@ -0,0 +1,69 @@
...
+      <vui-button class="danger danger-control" icon="exclamation-triangle" :params="{abort: 1}">Abort!</vui-button>
+    </template>
+    <template v-else>
+      <h4>Ingredients</h4>
+      <ul>
+        <li v-if="!ingredientsPresent">The microwave is empty!</li>
+        <li v-for="(amt, name) in cookingobjs" :key="name">
+          <vui-button push-state :params="{ eject: name }"> {{ name }} • {{ amt }}
```
Comment (Karolis2011) [inline vueui/src/components/view/cooking/microwave.vue]: Why am I still seeing push state still set after I scolded you for using it last time?

File: code/game/machinery/kitchen/microwave.dm (hunk 23 above)
Comment (JohnWildkins) [reply code/game/machinery/kitchen/microwave.dm]: Would that be significantly better? You warned me against altering UI data directly before.

Comment (Karolis2011) [reply code/game/machinery/kitchen/microwave.dm]: Well, might not be a huge difference, but if it happens a lot it might become an issue. Might be fine tho. Also directly editing UI object data is not recommended, but if its only way to get things to properly update more complex structures then it's okay. Like I am sure I am editing UI data inside record computers.

**Comment by JohnWildkins** [comment]
> 
> 
> Might better just do `max(this.min, min(this.max, this.value))` instead of `this.value` this would actually be better because it make it cap from bottom and from top. And might be faster than its.

Yeah, that looks better. Let me know if that looks good or if there's anything else.
